### ISSUE TITLE
Add plain SQL versions of select(String), groupBy(String), orderBy(String)

### ISSUE THREAD

**Issue by nkoterba** [jOOQ/jOOQ#5003]
So I noticed that one of the options of `DSL.condition()` allows creating a custom where clause using either a `SQL` object or a `String` object.  I'm assuming this is to allow custom, complex `WHERE` clauses that may be difficult to construct with the DSL or not known ahead of time.  Internally, I'm assuming that when the final SQL string or command is generated, it just inserts this `WHERE` clause into the correct place in the query string.

Does Jooq support custom clauses, again either as a `SQL` or `String` objects for the other parts of a query?  Namely, the select, orderBy, and groupBy clauses?  I don't see similar functionality for these parts of a query.

I recognize that `WHERE` clauses can be easily combined and chained so allowing a `condition` to be created with custom sql fits nicely into the builder pipeline.  The `from` method in Jooq DSL also supports a `SQL` or `String` object.

Since the `groupBy()`, `orderBy()`, and `select()` methods all force developers to "move on" to the next step, I would think that allowing custom SQL here would also be possible.

Of course, allowing users to define custom SQL means they could write incorrect syntax and/or mismatched columns (e.g. adding a groupBy clause that contains columns not in the select clause).  But such issues would hopefully appear quickly as the SQL processor in the datastore would throw errors.

In short, I think it would be useful (at least in our use cases) to be able to do:

``` java
dslContext
  .select("id, count(*)")
  .from(TABLE)
  .where(DSL.condition(<custom sql>))
  .groupBy(DSL.group(<custom sql>))
  .orderBy(TABLE.ID)

// OR

  dslContext
  .select(TABLE.NAME, TABLE.NAME.count())
  .from("table")
  .where(TABLE.ID.ge(20))
  .groupBy(TABLE.NAME)
  .orderBy(DSL.order(<custom sql>))
```

**Comment by lukaseder** [comment]
I suspect what you're looking for is [`DSL.field(String)`](http://www.jooq.org/javadoc/latest/org/jooq/impl/DSL.html#field-java.lang.String-) and related methods.

In jOOQ, this concept is called plain SQL, and it has a whole chapter in the manual:
http://www.jooq.org/doc/latest/manual/sql-building/plain-sql

Note that jOOQ won't allow you to change / add custom *clauses*, but only certain *expressions*, such as expressions of type `Condition`, `Table`, `Field`

I hope this helps,
Lukas

**Comment by nkoterba** [comment]
@lukaseder Thanks for the quick reply.  I am familiar with the sql-building for `Condition`, `Table`, and `Field`.

Unfortunately, that doesn't work well when the legacy aspects of our system pass say a `String` object for `select` or `orderBy` clauses (I know...not the best design...lots of issues including fat fingering mistakes as well as SQL injection -- though we don't directly take field input and blindly concatenate it into these strings).

To use the current available features, we would have to take a `ORDERBY` statement we have as a string:

`devicename desc, networkname desc, building_id desc`

and parse the string, creating a `Field` for each delimited item and then correctly detecting order by direction to allow us to write  (see below).  

```java

Collection<SortField<?>> sortFields = // calls some parse function which returns:

  // DSL.field("devicename").desc(),
  // DSL.field("networkname").desc(),
  // DSL.field("building_id").desc()

Query q = dslContext
.selectFrom(DEVICE)
.orderBy(sortFields);
```

Certainly doable but was hoping we could "transition" into Jooq by converting parts of our queries over time.   We'll find a good compromise until such functionality is potentially available.  Just wanted to make sure there wasn't something glaringly obvious we were missing.

**Comment by lukaseder** [comment]
I must admit, it had never occurred to me to allow for plain SQL to be used for an entire clause (even if `from(String)`, `where(String)`, `having(String)`, `connectBy(String)` already exist...

We'll implement that for jOOQ 3.8, it makes perfect sense from an API consistency point of view.

**Comment by lukaseder** [comment]
As far as this is concerned:

> Just wanted to make sure there wasn't something glaringly obvious we were missing.

You can still work around this lacking feature by specifying `DSL.field(String)` to `groupBy()` and `orderBy()`. In the `select()` clause, this probably won't work, as jOOQ will not be able to recognise the fact that your single `Field` in fact represents several fields. I suspect that will cause issues when executing the query.

You could work around this issue by implementing a `VisitListener` for this, but that would be as much work as cleanly splitting the fields in the client code, prior to passing the fields to `select()`

**Comment by nkoterba** [comment]
:+1: on the comments, followup, and suggestions.

We are going with the string splitting process in the sections where we need it for the time being.

I think there could be the following set of functions:
```java
select(String)
from(String)  // CURRENTLY EXISTS
where(String) // CURRENTLY EXISTS
groupBy(String)
having(String) // CURRENTLY EXISTS
orderBy(String)
connectBy(String) // CURRENTLY EXISTS
```

However, I noticed that Jooq's behavior differs on how it handles each of the above (at least based on Jooq comments and some simple exploration).  You allude to this in your comment about `select(String)`.

For example:
`from(String)` says that the String must evaluate to a table that can be found in the ResultSetMetaData.  That means writing something like `from("company INNER JOIN employee ON employee.company_id = company.id")` will fail or cause errors because that statement is a JOIN and no single table can be easily extracted form the ResultSetMetaData.

Whereas, `where(String)` can be any custom SQL string, e.g. `where("id = 3 AND name = 'bob'")`, and jooq is happy just inserting that into the final SQL string.

I haven't looked into the `having(String)` function or `connectBy(String)` function yet to see whether they behave more like `from` or `where` but I'm guessing it's more like `where`.

So if indeed this feature is implemented, I think there could be some confusion as to whether the provided "raw" SQL needs to be restricted in such ways that the Jooq internals can "process it" OR whether it indeed can be "any valid" SQL.

I recognize that moving towards any valid SQL somewhat defeats the intent of Jooq to avoid writing essentially String-based SQL statements.  For example, at the extreme Jooq could support the following:

```java
dslContext
  .select("id, name, company_name")
  .from("employee INNER JOIN company on employee.company_id = company.id")
  .where("id IN (1,2,3,4,5,6,7,8,9,10)")
  .groupBy("company_name")
  .having("count(id) > 2")
  .orderBy("company_name")
  .toSql();
```

**Comment by lukaseder** [comment]
> from(String) says that the String must evaluate to a table that can be found in the ResultSetMetaData. 

Hmm, where did you find that? I don't think that's correct. You can put anything in `from(String)`, including what you suggested.

> So if indeed this feature is implemented, I think there could be some confusion as to whether the provided "raw" SQL needs to be restricted in such ways that the Jooq internals can "process it" OR whether it indeed can be "any valid" SQL.

The challenge is really getting all the semantics of all the different clauses right. If in doubt, the worst thing is always `LIMIT .. OFFSET` emulation, e.g. for Oracle 11g and less. This becomes impossible, if the `SELECT` clause is unknown.

> I recognize that moving towards any valid SQL somewhat defeats the intent of Jooq to avoid writing essentially String-based SQL statements. For example, at the extreme Jooq could support the following: [...]

Yes, that would be a bit weird. At this point, you might as well stringify those `select()`, `from()` keywords and go all plain SQL :)

**Comment by nkoterba** [comment]
>> from(String) says that the String must evaluate to a table that can be found in the ResultSetMetaData.

> Hmm, where did you find that? I don't think that's correct. You can put anything in from(String), including what you suggested.

I apologize...it's on `DSL.table(String)`: https://github.com/jOOQ/jOOQ/blob/master/jOOQ/src/main/java/org/jooq/impl/DSL.java#L7759-L7790.

You're correct that `dslContext.select(...).from(String)` javadoc doesn't say anything to that affect.  I got the two confused.  I assumed the `DSL.table(String)` and `.from(String)` did essentially the same thing, but one returns a `Table` while the other returns a `SelectJoinStep`.  I guess I was considering the case where you might do:

`dslContext.selectFrom(DSL.table('company INNER JOIN employee on company.id = employee.company_id'))`

**Comment by lukaseder** [comment]
Hmm, thanks for following up. That's indeed a bit confusing (and wrong). I've registered an issue to fix the Javadoc: https://github.com/jOOQ/jOOQ/issues/5004.

This restriction was a nice idea, but we never implemented it this way, as a plain SQL table declaration cannot be taken out of its enclosing context to find out what fields are available.

Besides, as you may have noticed, `from(String)` simply delegates to `DSL.table(String)`, so the restriction isn't enforced, there.

**Comment by lukaseder** [comment]
I've been postponing this issue for a while now, based on the grounds that mainly the `select(String)` overload is clearly not where we want to go with the jOOQ API, for at least these 2 reasons:

- Users might mistakenly pass a `String` bind value to it, expecting it to be the same as `select(val(String))`, when it is in fact `select(field(String))`. This accident isn't unlikely, and it could produce bugs or SQLi vulnerabilities. Also, we would prevent any future API evolution where it is desirable to pass bind values to `select()`
- A central part of jOOQ's type safety is the type safe production of a `Record[N]` type (e.g. `Record2<Integer, String>`). Accepting a plain SQL projection goes against this. Your suggestion of putting multiple columns in a single column does, too. Once you're ready to do that heavy trickery, why not just write `dsl.resultQuery("SELECT a, b, c FROM t")`. What remaining benefit are you getting out of jOOQ's DSL?

All of the plain SQL templating API is meant as an escape hatch for when jOOQ really can't support the feature natively. Not as a general purpose feature that should be used widely. As such, it doesn't really deserve the prominent visibility it has today. It is much better to migrate query by query to jOOQ, rewriting the query to something idiomatic, rather than migrating every query but only towards using plain SQL.

So, I don't confirm this use-case you're describing. It's not something jOOQ would want to encourage. At the same time:

- The `from(String)` methods, etc. have been a bit of a mistake. It should *never* be possible to *accidentally* create a plain SQL template. The `DSL.field(String)` method is far too error prone, and I regret it. All of these plain SQL templating methods should at least accept the `SQL` type, which makes their intent very clear, and whose construction overhead is acceptable for the occasional escape hatch use-case: `field(sql("something(x, ?)", bind), dataType)`. Too verbose for general purpose usage, but perfectly acceptable for the intended purpose.
- Every plain SQL templating use-case can be implemented for every type of `QueryPart` via some additional, explicit wrappers. The convenience use-case isn't a primary use-case, and personally, I think that every example out there (e.g. on Stack Overflow) by a user who doesn't want to go all in on jOOQ *just yet* makes a bad impression of jOOQ, because the queries look weird and unclean, and if going all in is such an impediment now, why wouldn't it be, later? There's no guarantee that those unclean queries will ever be cleaned up, so I don't think that's a reasonable intermediate step. Worse. Those developers will then blame jOOQ for the unclean code, not the fact that they didn't go all in. There are hundreds of sloppy query builders out there who can do that and only that. But they add zero value, they just complicate things compared to a simple static String based native SQL query, whereas jOOQ's full power is achieved by using the code generator, dynamic SQL, and all the fancy features, including `MULTISET`, nested records, implicit join, and much more, see: https://blog.jooq.org/why-you-should-use-jooq-with-code-generation/. I've written that blog post because I was fed up with support cases where users didn't want to use code generation because of the extra setup costs (might do it "later"), only to write sloppy plain SQL jOOQ queries, running into edge cases and bugs that they would not have run into, had they used the code generator.

You, yourself, wrote:

> I recognize that moving towards any valid SQL somewhat defeats the intent of Jooq to avoid writing essentially String-based SQL statements. For example, at the extreme Jooq could support the following: [...]

And that's precisely what a lot of people end up doing already today. Not just people who want to migrate slowly, but people who build new applications with jOOQ. There are now 2 books on jOOQ (1 published, 1 soon to be published), and both authors explain query building with the plain SQL API first, the methodical approach being to postpone explaining the code generator. While I understand the rationale of postponing additional cognitive friction, it's really a pity, because many users will stick with what has worked for them, never upgrading to the ideal jOOQ usage. And the simplicity of the plain SQL API encourages this, which, in retrospect, was a big API design mistake.

As such, I'm more inclined to eventually deprecate some of the existing plain SQL templating convenience (not the underlying primitives, of course, just the convenience) in favour of a more code generation oriented, type safe jOOQ. The deprecation won't be done eagerly, but this is to make a final stance on this issue, which I'll now reject.
